File: CokeLee777/src/com/leetcode/greedy/ContiguousArray.java
package com.leetcode.greedy;

import java.util.HashMap;
import java.util.Map;

/**
 * https://leetcode.com/problems/contiguous-array/
 */
public class ContiguousArray {

    public static class Solution{

        public int findMaxLength(int[] nums) {
            // 배열안의 0인 원소들 -1로 초기화
            for (int i = 0; i < nums.length; i++) {
                if (nums[i] == 0) nums[i] = -1;
            }
            // Key: sum, Value: index
            Map<Integer, Integer> sumToIndex = new HashMap<>();
            /**
             * ====== Key Point ======
             * index: -1 의 합을 0으로 초기화 -> 첫번째 인덱스부터의 길이를 비교하기 위함
             */
            sumToIndex.put(0, -1);

            int sum = 0, max = 0;
            for (int i = 0; i < nums.length; i++) {
                //현재값 더하기
                sum += nums[i];
                //현재 인덱스까지의 합과 동일한 값이 존재한다면 최대길이 비교
                if (sumToIndex.containsKey(sum)) {
                    max = Math.max(max, i - sumToIndex.get(sum));
                }
                //아니라면 현재 인덱스까지의 합 map 에 삽입
                else {
                    sumToIndex.put(sum, i);
                }
            }

            return max;
        }
    }

}

File: dongdaang/greedy/Contiguous_Array.py
# ############시간초과 초드...#############

# class Solution:
#     def findMaxLength(self, nums):
#         s = sum(nums)
#         # if s == len(nums) // 2:
#         #     return len(nums)
#         i = 0
#         while True:
#             tmp = []
#             for j in range(i+1):
#                 tmp = nums[:j] + nums[len(nums)-i+j:]
#                 if s - sum(tmp) == (len(nums) - i) / 2:
#                     return len(nums) - i
#             i += 1
            
# a = Solution()
# print(a.findMaxLength([0,1,0,0,0,1,1,1,0,1,1]))

##############솔루션 확인###############

class Solution(object):
    def findMaxLength(self, nums):
        count = 0
        max_length=0
        table = {0: 0}
        for index, num in enumerate(nums, 1):
            if num == 0:
                count -= 1
            else:
                count += 1
            
            if count in table:
                max_length = max(max_length, index - table[count])
            else:
                table[count] = index
        
        return max_length